src/DataAPI.jl

```diff
@@ -287,4 +287,158 @@ using a `sink` function to materialize the table.
 """
 function allcombinations end
 
+const STYLE_INFO = """
+One of the uses of the metadata `style` is decision
+how the metadata should be propagated when `x` is transformed. This interface
+defines the `:none` style that indicates that metadata should not be propagated
+under any operations (it is only preserved when a copy of the source table is
+performed). All types supporting metadata allow at least this style.
+"""
+
+const COL_INFO = """
+`col` must have a type that is supported by table `x` for column indexing.
+Following the Tables.jl contract `Symbol` and `Int` are always allowed.
+Passing `col` that is not a column of `x` throws an error.
+"""
+
+"""
+    metadata(x, key::AbstractString; style::Bool=false)
+
+Return metadata value associated with object `x` for key `key`.
+If `x` does not support metadata throw `ArgumentError`.
+If `x` supports metadata, but does not have a mapping for `key` throw
+`KeyError`.
+
+If `style=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored
+for the `key`.
+
+$STYLE_INFO
+"""
+metadata(::T, ::AbstractString; style::Bool=false) where {T} =
+    throw(ArgumentError("Objects of type $T do not support getting metadata"))
+
+"""
+    metadatakeys(x)
+
+Return an iterator of metadata keys for which `metadata(x, key)` returns a
+metadata value. If `x` does not support metadata return `()`.
+"""
+metadatakeys(::Any) = ()
+
+"""
+    metadata!(x, key::AbstractString, value; style)
```

Comment:
Thinking about it, maybe the syntax would be more natural as `metadata!(x, key => value)`? That would allow extending this in the future to pass multiple pairs if it appears to be convenient.

A counter-argument is that `setindex!` doesn't use that syntax, but it's almost never called that way since `x[key] = value` is nicer. Of course both syntaxes could be allowed as they are not ambiguous (we will probably never allow keys to be pairs).

Reply:
As a first reaction it makes sense.

My only reservation was that in DataFrames.jl `=>` is used for operation specification language, so we would have yet a third way to interpret `=>` there.

The question is how would it look for `colmetadata!`?
```
colmetadata!(x, col, key => value; style=style)
```
(which does not look that nice)

Also note that `metadata!(x, key => value)` would not be allowed, you would need to write `metadata!(x, key => value; style=style)`.

In summary - I would keep the things as they are here and consider what we design here a low-level API.

I assume that the extra new package planned (tentatively named TableMetadataTools.jl) will provide convenient high-level functions. In practice I even expect that if we define there:
```
caption!(table, str) = metadata!(table, "caption", str, style=:note)
caption(table) = metadata(table, "caption")
label!(table, str) = colmetadata!(table, col, "label", str, style=:note)
label(table, col) = colmetadata(table, col, "caption")
```
this will cover 95% of use cases of metadata in practice.

In summary - I propose to discuss a convenience high-level API in TableMetadataTools.jl, as I expect that in that package we will drop the requirement to specify `style` which we have in low-level API, as in high level API all styles will be `:note`.